### PR TITLE
Johanna/2908 react 18 upgrade

### PR DESCRIPTION
**Pull request by johanna-skylight**
# FRONTEND PULL REQUEST

## Related Issue

Implements #2908 

## Changes Proposed

Below I have listed the common changes that you will see across this PR

-  Added additional element checks in spec 06 and 07. Tests were failing because cypress was trying to get hold of an element when the UI was not done rerendering. 

- Storybook dependency and setup update. Storybook setup in general was updated. I ran the storybook wizard which took care of rewriting the setup and updating its dependencies with the latest dependencies and versions.

- Added children as an explicit prop in multiple components. Moving forward children is not an intrinsic prop so it needs to be explicitly added for the prop type of a component.

- Focus on error for multiple forms. The logic to focus on the first form field that reports an error after a form validation needed to be refactored as the former logic was triggering prior to the render that sets the DOM elements with the error states, hence never focusing on any element. 

-  Added async/await to userevent methods across jest test cases. Tests started to fail due to assertions getting triggered while the mocked user event was not done updating the UI.

- Added delays to endpoint mocks so jest can validate loading masks behaviors.

- Refactoring of the managed organization component for form validation and error display. The former implementation was relying on independent states to be available and updated at the same time in order to properly display the error toast contents. This was causing the toast to randomly display incomplete information if the logic that triggered the toast kicked before the other states were ready.

- Added type casting whenever font awesome svg elements are used.

- Added initialization flags in multiple components that are part of the account creation flow. Effects that were meant to be only called once during component initialization were being triggered multiple times. These effects were normally performing token initializations with okta endpoints causing it to invalidate the authentication flow while our cypress spec was still performing other steps of the process.

- Refactoring of the navigation functionality in the MfaTop component. The logic implemented in this flow to move to the next step of the account creation was relying on state updates to trigger the navigation (user clicks button-> we update some internal state -> the state update triggers the router navigation). This was breaking the component during cypress testing. Removing the unnecessary usage of a state and cutting the re rendering by just triggering the router navigation upon the click of the user fixed the issue. 

- Refactoring of passwordForm component for the strength validation of a password. Similar to the issue mentioned in the managed organization component. This implementation was relying on independent states to be available and updated at the same time in order to properly unlock other component logic. Two states were bundle in one and a type was added for implementation clarity.
 
-  Refactoring of index.tsx. In the current implementation this file is performing the role of root element for our application as well as handling the injection of the root element into the html dom with the react-dom methods. The ideal implementation in a react application is that app.tsx is the root element of the application and index.tsx only handles the injection of the element to the HTML DOM. The uncommon current implementation was also causing the index.test.tsx to fail after the migration to the new createRoot method was done. 

            So three things happened: 
            1- index.tsx logic was cut to only cover the injection to the DOM 
            2- All the other logic that set up the main routes, integrated the store, integrated apollo, integrated telemetry and other global providers was migrated to app.tsx.
            3- The current app.tsx that was hosting the reporting logic was renamed to reportingApp.tsx.

## Additional Information
Why the dependency upgrades and the code refactoring I saw a decrease of 10 sec for in the jest testing. In my local the total time went from 39s to 29s. I am curious to see if someone else sees this improvement.

This PR is really prone to conflicts whenever any new frontend code makes it to main so you will see me constantly pushing changes to fix this until this code gets merged.

## Testing
Regression testing. I have manually tested the app in dev5 but any general regression testing is very appreciated!

### CODE REVIEW

File: frontend/src/patientApp/timeOfTest/ToS.tsx
Comment (emyl3) [inline frontend/src/patientApp/timeOfTest/ToS.tsx:26]: out of curiosity, is there a reason we need to wrap this in `<>` now?

**Comment by emyl3** [comment]
> I saw a decrease of 10 sec for in the jest testing. In my local the total time went from 39s to 29s. I am curious to see if someone else sees this improvement.
I also saw improvements in how fast the tests ran. It went from ~48s to ~33s on your branch!! 🎉 nice!!

File: frontend/src/App.tsx
Comment (emyl3) [inline frontend/src/App.tsx:21]: thank you for refactoring the old index.tsx file and explaining why that was needed!! 👍

File: frontend/src/app/patients/EditPatient.test.tsx
Comment (emyl3) [inline frontend/src/app/patients/EditPatient.test.tsx:364]: nit: we can remove this commented out line :D

**Review by emyl3 — COMMENTED** [review]
reviewed a little under half of the files 🥲 Thank you for all your work on this! 😭 🙌 Will pick up reviewing the rest of it later today. 

Do you think it would be helpful to schedule a short group regression testing session (even if it is with just the G&E devs)? 🤔 

I've done some regression testing locally and so far things look good -- will also test once it's deployed to a lower environment as well!! Thank you so much, Johanna!!! 😸

File: frontend/src/app/patients/EditPatient.test.tsx
Comment (johanna-skylight) [reply frontend/src/app/patients/EditPatient.test.tsx:364]: thanks for catching this!

File: frontend/src/patientApp/timeOfTest/ToS.tsx
Comment (johanna-skylight) [reply frontend/src/patientApp/timeOfTest/ToS.tsx:26]: Typescript was giving me errors about passing more than one node to the upper component, so I just added a fragment to get rid of the error

File: frontend/src/app/Settings/Facility/FacilityForm.tsx
Comment (emyl3) [inline frontend/src/app/Settings/Facility/FacilityForm.tsx:111]: I think we also need `errors.deviceTypes` here in case a user gets rid of all the devices in their facility and clicks the save button at the top of the form

File: frontend/src/app/testQueue/addToQueue/SearchResults.stories.tsx
Comment (emyl3) [inline frontend/src/app/testQueue/addToQueue/SearchResults.stories.tsx:24]: wondering if we should put this in a `types.d.ts` file to call that instead?

**Review by emyl3 — COMMENTED** [review]
Down to the last 30 or so files!!! added a couple more comments!! great work, johanna! Will finish up tomorrow morning 🌻

File: frontend/src/app/commonComponents/SingleFileInput.tsx
Comment (emyl3) [inline frontend/src/app/commonComponents/SingleFileInput.tsx:48]: I think this should be just "Drag file here or choose from folder to change file"  (file singular) since we only support one file.

**Review by emyl3 — COMMENTED** [review]
Left another comment! the new file input functioned correctly! thank you for fixing that johanna! Overall codewise looks good. Did some more regression testing locally and looked good. Looking forward to doing the regression testing when we set that up!

File: frontend/package.json
Comment (emyl3) [inline frontend/package.json:23]: Can we set `@types/testing-library__jest-dom` to `"^5.14.5"` to match the update we are doing to `@testing-library/jest-dom` on line 199. 🤔 

haha only noticed it due to this dependabot PR: https://github.com/CDCgov/prime-simplereport/pull/4918

File: frontend/src/app/commonComponents/SingleFileInput.tsx
Comment (johanna-skylight) [reply frontend/src/app/commonComponents/SingleFileInput.tsx]: Totally on point, thanks for the suggestion :)

File: frontend/src/app/testQueue/addToQueue/SearchResults.stories.tsx
Comment (johanna-skylight) [reply frontend/src/app/testQueue/addToQueue/SearchResults.stories.tsx]: I have moved this type to a global setup in react-app-env.d.ts

File: frontend/src/app/signUp/IdentityVerification/PersonalDetailsForm.tsx
Comment (emmastephenson) [inline frontend/src/app/signUp/IdentityVerification/PersonalDetailsForm.tsx:104]: I saw this method for focusing error on another component as well - is it potentially worth extracting the logic into a helper method, or do you just think that would make everything more complicated? I'm specifically talking about the body of the `if` statement.

File: frontend/src/app/signUp/IdentityVerification/QuestionsForm.test.tsx
Comment (emmastephenson) [inline frontend/src/app/signUp/IdentityVerification/QuestionsForm.test.tsx:37]: Nice, this is much more readable/standardized with the fake timer

**Review by emmastephenson — COMMENTED** [review]
I still have 20-30 more files to review, but looks great so far! Thanks for all your effort on this, Johanna! 👏

**Review by emmastephenson — COMMENTED** [review]
Code changes all look great! Just holding off on final approval until after we get the chance to bug bash this.

Thanks for all your awesome work here, Johanna! 🙏

File: frontend/src/app/Settings/Facility/FacilityForm.tsx
Comment (johanna-skylight) [reply frontend/src/app/Settings/Facility/FacilityForm.tsx:112]: Thanks for pointing this out! I was hoping someone more familiar with the logic would chime in with suggestions :D

File: frontend/src/App.tsx
Comment (johanna-skylight) [reply frontend/src/App.tsx:46]: yeah, plus the library to capture the logs changed the way to access the user session Id so that logic needs to be checked.

File: frontend/src/App.test.tsx
Comment (johanna-skylight) [reply frontend/src/App.test.tsx:12]: true!

File: frontend/src/app/signUp/IdentityVerification/PersonalDetailsForm.tsx
Comment (johanna-skylight) [reply frontend/src/app/signUp/IdentityVerification/PersonalDetailsForm.tsx:97]: Good idea!

File: frontend/src/app/patients/AddPatient.test.tsx
Comment (johanna-skylight) [reply frontend/src/app/patients/AddPatient.test.tsx]: I have moved the loading mask checks in favor of avoiding the introduction of a delay in the response mock

File: frontend/src/app/ReportingApp.test.tsx
Comment (johanna-skylight) [reply frontend/src/app/ReportingApp.test.tsx:223]: the native jest fake timers started to have a weird behavior with the upgrade. This PR is not only upgrading react but jest and react testing library as well. I am not sure which of those dependencies introduced the issues but basically if you are not properly wrapping the fake timers methods in act and setting them up and cleaning them up correctly in the test file they start to introduce leaks or infinite loops in the testing.

If you need to test time moving forwards (like the test timer in the test card) for all means use the fake timers and deal with its complexity, but if the only thing you need is to set a fake date in the system then the MockDate library is so much easier and cleaner to use.

File: frontend/src/app/Settings/Facility/ManageFacilitiesContainer.test.tsx
Comment (fzhao99) [inline frontend/src/app/Settings/Facility/ManageFacilitiesContainer.test.tsx:73]: 🎉

File: frontend/src/app/accountCreation/AccountCreationApp.tsx
Comment (fzhao99) [inline frontend/src/app/accountCreation/AccountCreationApp.tsx:77]: question: would we ever get an unwanted rerender adding these values to the dependency array? Notice that we're changing this useEffect from being a "run once on app load" to depending on these values in React Router changing.

Don't have the context for RR at all though, so totally could be missing something!

File: frontend/src/app/analytics/Analytics.test.tsx
Comment (fzhao99) [inline frontend/src/app/analytics/Analytics.test.tsx:291]: question: is there a reason we want the fireEvent API here rather than the higher-level userEvent? Guessing that the test was flaking for some reason with `userEvent.type`, but wanted to confirm / add context here since [the docs call out this difference explicitly](https://testing-library.com/docs/dom-testing-library/api-events/)

File: frontend/src/app/patients/AddPatient.test.tsx
Comment (fzhao99) [inline frontend/src/app/patients/AddPatient.test.tsx:51]: ditto to comment above on why we're using fireEvent over userEvent

**Comment by fzhao99** [comment]
about halfway through, but based on what i've seen, kudos on the tremendous effort that went into this Johanna! Excited to get this across the finish line

File: frontend/src/app/analytics/Analytics.test.tsx
Comment (johanna-skylight) [reply frontend/src/app/analytics/Analytics.test.tsx:291]: user event library is not working correctly for inputs of type date. Sometimes the inputs do not get updated.  Changing the type event back to the fireEvent library makes the tests less flaky

File: frontend/src/app/patients/AddPatient.test.tsx
Comment (johanna-skylight) [reply frontend/src/app/patients/AddPatient.test.tsx:51]: same answer as before!

File: frontend/src/app/accountCreation/AccountCreationApp.tsx
Comment (johanna-skylight) [reply frontend/src/app/accountCreation/AccountCreationApp.tsx:77]: a few things happened behind this code change. 
1- Even when the dependency array was empty this logic was triggering more than once during cypress testing. So to make sure this only triggers once an initialization flag has been added.
2- Given that the logic to activate the token is now guarded behind the flag I added the dependencies to the array as this is actually the way is supposed to happen. Pathname as primitive value and navigate as ref passed as a hook  are not triggering weird renders.

File: frontend/src/app/analytics/Analytics.test.tsx
Comment (fzhao99) [reply frontend/src/app/analytics/Analytics.test.tsx:291]: I did come across [this comment](https://github.com/testing-library/user-event/issues/399#issuecomment-815664027) in the RTL repo about `userEvent` needing to be `.clear()`-ed first for date inputs. Tested it locally and it did seem to fix flakeage.

No need to fix it for this PR, but thought I'd leave it here for the record.

**Review by emmastephenson — DISMISSED** [review]
Changes look great! I just checked Azure and the telemetry all looks good 👍 Thanks for your hard work here, I think we're good to merge these changes! 

One note is that if the changes require any local dev updates (running `yarn install` or what have you), please make sure to note that in the eng Slack channel once it's merged.

**Review by emyl3 — APPROVED** [review]
LGTM!!! Thank you for all your work on this, johanna!!! 😭 🙌

**Review by emyl3 — APPROVED** [review]
so sorry for having you resolve more merge conflicts with my merge 😭 Thank you for doing that!!! I will merge no more code until this is in 😛

**Review by emmastephenson — APPROVED** [review]
LGTM!

**Pull request merged**
